sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPluginInternal.scala

```diff
@@ -412,20 +418,34 @@ object ScalaJSPluginInternal {
         ((crossTarget in packageScalaJSLauncher).value /
             ((moduleName in packageScalaJSLauncher).value + "-launcher.js")),
 
-      skip in packageScalaJSLauncher := !persistLauncher.value,
+      skip in packageScalaJSLauncher := {
+        val value = !persistLauncher.value
+        if (!value && (scalaJSModuleKind.value != ModuleKind.NoModule)) {
+          throw new IllegalArgumentException(
+              "persistLauncher := true is not compatible with emitting JavaScript modules")
+        }
+        value
+      },
 
       packageScalaJSLauncher <<= Def.taskDyn {
-        if ((skip in packageScalaJSLauncher).value)
-          Def.task(Attributed.blank((artifactPath in packageScalaJSLauncher).value))
-        else Def.task {
-          mainClass.value map { mainCl =>
-            val file = (artifactPath in packageScalaJSLauncher).value
-            IO.write(file, launcherContent(mainCl), Charset.forName("UTF-8"))
-
-            // Attach the name of the main class used, (ab?)using the name key
-            Attributed(file)(AttributeMap.empty.put(name.key, mainCl))
-          } getOrElse {
-            sys.error("Cannot write launcher file, since there is no or multiple mainClasses")
+        if ((skip in packageScalaJSLauncher).value) {
+          Def.task {
+            Attributed.blank((artifactPath in packageScalaJSLauncher).value)
+          }
+        } else {
+          Def.task {
+            mainClass.value map { mainCl =>
+              val file = (artifactPath in packageScalaJSLauncher).value
+              val moduleKind = scalaJSModuleKind.value
+              IO.write(file,
+                  launcherContent(mainCl, moduleKind, None),
```

Comment:
Can you pass in `moduleKind.NoModule` (or assert it) to make it explicit that this does not work with modules?